dbt_subprojects/daily_spellbook/models/_sector/addresses/addresses_info.sql

```diff
@@ -31,240 +30,198 @@
+        , am.tokens_sent_count
+        , am.tokens_sent_tx_count
+        , am.first_transfer_block_time
+        , am.last_transfer_block_time
+        , am.first_received_block_number
+        , am.last_received_block_number
+        , am.first_sent_block_number
+        , am.last_sent_block_number
+        , am.received_volume_usd
+        , am.sent_volume_usd
+        , am.first_tx_block_time
+        , am.last_tx_block_time
+        , am.first_tx_block_number
+        , am.last_tx_block_number
+        , map_from_entries(array[
+            ('last_seen', CAST(am.last_seen AS varchar))
+            , ('last_seen_block', CAST(am.last_seen_block AS varchar))
+            , ('executed_tx_count', CAST(am.executed_tx_count AS varchar))
+            , ('is_smart_contract', CAST(am.is_smart_contract AS varchar))
+            , ('tokens_sent_count', CAST(am.tokens_sent_count AS varchar))
+            , ('tokens_received_count', CAST(am.tokens_received_count AS varchar))
+            ]) AS chain_stats
+        , am.last_seen
+        , am.last_seen_block
+        FROM {{ addresses_model[1] }} am
+        LEFT JOIN {{ this }} t ON am.address = t.address
+            AND (((contains(t.blockchains, am.blockchain) = FALSE))
+            OR (CAST(chain_stats['{{ addresses_model[0] }}']['last_seen_block'] AS bigint) > t.last_seen_block))
+        WHERE {{incremental_predicate('am.last_seen')}}
+        {% if not loop.last %}
+        UNION ALL
+        {% endif %}
+        {% endfor %}
+        )
+    GROUP BY address
+    )
+
+SELECT nd.address
+, array_union(t.blockchains, nd.blockchains) AS blockchains
+, t.executed_tx_count+nd.executed_tx_count AS executed_tx_count
+, GREATEST(t.max_nonce, nd.max_nonce) AS max_nonce
+, CASE WHEN GREATEST(t.max_nonce, nd.max_nonce) = t.max_nonce THEN t.max_nonce_blockchain
+    ELSE nd.max_nonce_blockchain
+    END AS max_nonce_blockchain
+, GREATEST(t.is_smart_contract, nd.is_smart_contract) AS is_smart_contract
+, array_union(t.smart_contract_blockchains, nd.smart_contract_blockchains) AS smart_contract_blockchains
+, COALESCE(nd.namespace, t.namespace) AS namespace
+, COALESCE(nd.name, t.name) AS name
+, CASE WHEN LEAST(t.first_funded_by_block_time, nd.first_funded_by_block_time) = t.first_funded_by_block_time THEN t.first_funded_by
+    ELSE nd.first_funded_by END AS first_funded_by
+, CASE WHEN LEAST(t.first_funded_by_block_time, nd.first_funded_by_block_time) = t.first_funded_by_block_time THEN t.first_funded_blockchain
+    ELSE nd.first_funded_blockchain END AS first_funded_blockchain
+, LEAST(t.first_funded_by_block_time, nd.first_funded_by_block_time) AS first_funded_by_block_time
+, t.tokens_received_count+nd.tokens_received_count AS tokens_received_count
+, t.tokens_received_tx_count+nd.tokens_received_tx_count AS tokens_received_tx_count
+, t.tokens_sent_count+nd.tokens_sent_count AS tokens_sent_count
+, t.tokens_sent_tx_count+nd.tokens_sent_tx_count AS tokens_sent_tx_count
```

Comment:
Adding up the old (t) and new (nd) data is not correct and will result in double counting transactions and volume. 

When an address is already in the dataset, and it is updated with new data. 
The old data will contain the sum of all history up to 'last_update' and the the new data will contain the sum of all history up to 'new_update'. We would need to get only the part of the sum that is not yet accounted for, which is just not feasible to get in this incremental way.  

I will continue with making the model a table, and we can look to get it in that way.
This will result in the cross-chain model being fully refreshed each day. We'll have to monitor performance in prod if it's feasible to do it that way.

Reply:
> Adding up the old (t) and new (nd) data is not correct and will result in double counting transactions and volume.
> 
> When an address is already in the dataset, and it is updated with new data. The old data will contain the sum of all history up to 'last_update' and the new data will contain the sum of all history up to 'new_update'. We would need to get only the part of the sum that is not yet accounted for, which is just not feasible to get in this incremental way.
> 
> I will continue with making the model a table, and we can look to get it in that way. This will result in the cross-chain model being fully refreshed each day. We'll have to monitor performance in prod if it's feasible to do it that way.

But wdym, that's exactly what I'm doing here with this join. On incremental runs it uses this join to only gather new data either where there was no data for that address+chain combo, and if there was it only does from the last recorded `block_number` onwards (excluding the `block_number` itself to not double count its data). This is the whole reason behind the mapping, so that I can have access to that `block_number` value per chain. So summing up should work properly

<img width="912" alt="Screenshot 2024-10-17 at 13 22 14" src="https://github.com/user-attachments/assets/315dd85e-6a7f-4f10-aa5d-219784e22a3f">

Reply:
You are correctly getting only the addresses that have received an update.
But the 'tokens_received_tx_count` value in the chain specific model will still have the total sum for that address, not just the count of the last incremental window.